File: scripts/create-data-url.R
#!/usr/bin/env Rscript
#' Generate self-contained iframe HTML with embedded audio for R Markdown/Quarto
#'
#' Command-line usage:
#'   Rscript create-data-url.R audio.wav
#'   Rscript create-data-url.R audio.wav "pitch,formants,hnr"
#'   Rscript create-data-url.R audio.wav "pitch,formants" "https://mysite.com/viewer.html"
#'
#' Usage in R Markdown/Quarto:
#'   source("scripts/create-data-url.R")
#'   html <- create_embedded_viewer("audio.wav", overlays = "pitch,formants,hnr")
#'   htmltools::HTML(html)

create_embedded_viewer <- function(audio_path, overlays = "pitch,formants", viewer_url = "viewer.html") {
  if (!requireNamespace("base64enc", quietly = TRUE)) {
    stop("Package 'base64enc' required. Install with: install.packages('base64enc')")
  }

  # Check file exists
  if (!file.exists(audio_path)) {
    stop(paste("Audio file not found:", audio_path))
  }

  # Check file size
  size_mb <- file.info(audio_path)$size / (1024 * 1024)
  if (size_mb > 1.5) {
    warning(sprintf("File is %.1fMB. Data URLs are limited to ~1.5MB. Consider using a remote URL.", size_mb))
  }

  # Read and encode
  audio_data <- readBin(audio_path, "raw", file.info(audio_path)$size)
  b64 <- base64enc::base64encode(audio_data)

  # Create data URL
  data_url <- paste0("data:audio/wav;base64,", b64)

  # URL encode
  encoded <- URLencode(data_url, reserved = TRUE)

  # Create iframe
  iframe <- sprintf(
    '<iframe
  src="%s?audio=%s&overlays=%s"
  width="100%%"
  height="600"
  frameborder="0"
  style="border: 1px solid #ddd; border-radius: 4px;">
</iframe>',
    viewer_url,
    encoded,
    overlays
  )

  return(iframe)
}

# Command-line interface (only runs when executed as a script, not when sourced)
if (sys.nframe() == 0) {
  args <- commandArgs(trailingOnly = TRUE)

  if (length(args) < 1) {
    cat("Usage: Rscript create-data-url.R <audio-file> [overlays] [viewer-url]\n")
    cat("Example: Rscript create-data-url.R audio.wav 'pitch,formants,hnr' 'viewer.html'\n")
    quit(status = 1)
  }

  audio_path <- args[1]
  overlays <- if (length(args) >= 2) args[2] else "pitch,formants"
  viewer_url <- if (length(args) >= 3) args[3] else "viewer.html"

  tryCatch({
    html <- create_embedded_viewer(audio_path, overlays, viewer_url)
    cat(html, "\n")
  }, error = function(e) {
    cat("Error:", conditionMessage(e), "\n", file = stderr())
    quit(status = 1)
  })
}

File: scripts/create-data-url.py
#!/usr/bin/env python3
"""
Generate self-contained iframe HTML with embedded audio for Jupyter/Quarto.

Usage:
    python create-data-url.py audio.wav
    python create-data-url.py audio.wav --overlays pitch,formants,hnr
    python create-data-url.py audio.wav --overlays all --viewer-url https://mysite.com/viewer.html

IMPORTANT: To view the generated HTML locally, you must serve it over HTTP:
    python -m http.server 8000
    # Or use: python scripts/serve-quarto.py

Browsers block file:// iframes for security, so double-clicking the HTML won't work.
"""

import base64
import sys
from urllib.parse import quote
from pathlib import Path

def create_embedded_viewer(audio_path, overlays="pitch,formants", viewer_url="./ozen-web/viewer.html"):
    """Create self-contained iframe HTML with embedded audio."""
    audio_file = Path(audio_path)

    if not audio_file.exists():
        raise FileNotFoundError(f"Audio file not found: {audio_path}")

    # Check file size
    size_mb = audio_file.stat().st_size / (1024 * 1024)
    if size_mb > 1.5:
        print(f"Warning: File is {size_mb:.1f}MB. Data URLs are limited to ~1.5MB.", file=sys.stderr)
        print("Consider using a remote URL instead.", file=sys.stderr)

    # Read and encode
    with open(audio_file, 'rb') as f:
        audio_data = f.read()

    b64 = base64.b64encode(audio_data).decode('ascii')
    data_url = f"data:audio/wav;base64,{b64}"

    # URL encode
    encoded = quote(data_url, safe='')

    # Create iframe
    iframe = f'''<iframe
  src="{viewer_url}?audio={encoded}&overlays={overlays}"
  width="100%"
  height="600"
  frameborder="0"
  style="border: 1px solid #ddd; border-radius: 4px;">
</iframe>'''

    return iframe

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python create-data-url.py <audio-file> [--overlays pitch,formants] [--viewer-url ./ozen-web/viewer.html]")
        sys.exit(1)

    audio_path = sys.argv[1]
    overlays = "pitch,formants"
    viewer_url = "viewer.html"

    # Parse optional arguments
    i = 2
    while i < len(sys.argv):
        if sys.argv[i] == "--overlays" and i + 1 < len(sys.argv):
            overlays = sys.argv[i + 1]
            i += 2
        elif sys.argv[i] == "--viewer-url" and i + 1 < len(sys.argv):
            viewer_url = sys.argv[i + 1]
            i += 2
        else:
            print(f"Unknown argument: {sys.argv[i]}", file=sys.stderr)
            sys.exit(1)

    try:
        html = create_embedded_viewer(audio_path, overlays, viewer_url)
        print(html)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
